client/mysqltest.cc

```diff
@@ -3893,16 +4135,88 @@ void do_copy_file(struct st_command *command)
+    used and overwrites the file, we should allow chown as well.
+  */
+  recursive= 0;
+  do_not_overwrite= MY_DONT_OVERWRITE_FILE | MY_HOLD_ORIGINAL_MODES;
+  if (ds_options.str[0] == '-')
+    for (uint i= 1; i < ds_options.length; i++)
+      if (ds_options.str[i] == 'r')
+        recursive= 1;
+      else if (ds_options.str[i] == 'f')
+        do_not_overwrite= 0;
+      else
+      {
+        error= 1;
+        goto end;
+      }
+  else if (ds_options.length != 0)
+  {
+    error= 1;
+    goto end;
+  }
 
-  DBUG_PRINT("info", ("Copy %s to %s", ds_from_file.str, ds_to_file.str));
-  /* MY_HOLD_ORIGINAL_MODES prevents attempts to chown the file */
-  error= (my_copy(ds_from_file.str, ds_to_file.str,
-                  MYF(MY_DONT_OVERWRITE_FILE | MY_WME | MY_HOLD_ORIGINAL_MODES)) != 0);
+  /*
+    Throw an error if destination path is invalid, or source and destination
+    paths are same.
+  */
+  if (bad_path(ds_destination.str) ||
+      !strcmp(ds_source.str, ds_destination.str))
+  {
+    error= 1;
+  }
+  else if (my_stat(ds_source.str, &src_stat_buff, MYF(0)))
+  {
+    /* For source as regular file, check if destination is also a regular file */
+    if (MY_S_ISREG(src_stat_buff.st_mode))
+    {
+      /* If destination is a directory, append the file name */
+      if (my_stat(ds_destination.str, &dst_stat_buff, MYF(0)) && MY_S_ISDIR(dst_stat_buff.st_mode))
+      {
+        const char *wild= my_basename(ds_source.str);
+        char dir_separator[2]= {FN_LIBCHAR, 0};
+        dynstr_append_mem(&ds_destination, dir_separator, strlen(dir_separator));
+        dynstr_append_mem(&ds_destination, wild, strlen(wild));
+
+      }
+      DBUG_PRINT("info", ("Copy %s to %s", ds_source.str, ds_destination.str));
+      error= (my_copy(ds_source.str, ds_destination.str,
+              MYF(do_not_overwrite | MY_WME)) != 0);
+    }
+    /* For directory, copy if recursive is enabled */
+    else if (recursive)
+    {
+      DBUG_PRINT("info", ("Recursive copy files of %s to %s",
+                          ds_source.str, ds_destination.str));
+      DBUG_PRINT("info", ("listing directory: %s", ds_source.str));
+      error= recursive_copy(&ds_source, &ds_destination, do_not_overwrite, true);
+    }
+    else
+    {
+      error= 1;
+    }
+  }
+  else
+  {
+    /*
+      Cannot get file stat. Could be a wildcard pattern. Extract the basename
+      as wildcard pattern.  If it is not a valid wildcard pattern,
+      wildcard_copy() will return error.
+    */
+    const char *wild= my_basename(ds_source.str);
+    init_dynamic_string(&ds_wild, wild, strlen(wild) + 1, 0);
+    dynstr_trunc(&ds_source, ds_wild.length+1);
+    error= wildcard_copy(&ds_source, &ds_destination, &ds_wild, recursive, do_not_overwrite);
+  }
+
+end:
   handle_command_error(command, error, my_errno);
-  dynstr_free(&ds_from_file);
-  dynstr_free(&ds_to_file);
+  dynstr_free(&ds_source);
+  dynstr_free(&ds_destination);
+  dynstr_free(&ds_wild);
```

Comment:
this only works if ds_wild is always properly initialized.
move dynstr_free(&ds_wild) above, put it under error=wildcard_copy().